# Advanced Pandas techniques to boost performance!

https://engineering.upside.com/a-beginners-guide-to-optimizing-pandas-code-for-speed-c09ef2c6a4d6

Pycon Presentation - https://www.youtube.com/watch?v=HN5d490_KKk

We'll test and compare different methods of applying functions to pandas dataframes.
The function we'll be testing is the Haversine Distance formular which takes the cordinates of two points (lattitude and longitude) and calculates the straight line distance

In [1]:
import numpy as np

# Define a basic Haversine distance formula
def haversine(lat1, lon1, lat2, lon2):
    MILES = 3959
    lat1, lon1, lat2, lon2 = map(np.deg2rad, [lat1, lon1, lat2, lon2])
    dlat = lat2 - lat1 
    dlon = lon2 - lon1 
    a = np.sin(dlat/2)**2 + np.cos(lat1) * np.cos(lat2) * np.sin(dlon/2)**2
    c = 2 * np.arcsin(np.sqrt(a)) 
    total_miles = MILES * c
    return total_miles

# Let's loop though a dataframe of cordinates and return a series with these distences

In [2]:
# Define a function to manually loop over all rows and return a series of distances
def haversine_looping(df):
    distance_list = []
    for i in range(0, len(df)):
        d = haversine(40.671, -73.985, df.iloc[i]['latitude'], df.iloc[i]['longitude'])
        distance_list.append(d)
    return distance_list

In [3]:
import pandas as pd

file_name = "https://raw.githubusercontent.com/rajeevratan84/datascienceforbusiness/master/new_york_hotels.csv"
df = pd.read_csv(file_name, encoding = "ISO-8859-1")
df.head()

ean_hotel_id                                     name  ... high_rate  low_rate
0        269955       Hilton Garden Inn Albany/SUNY Area  ...  154.0272  124.0216
1        113431     Courtyard by Marriott Albany Thruway  ...  179.0100  134.0000
2        108151                    Radisson Hotel Albany  ...  134.1700   84.1600
3        254756  Hilton Garden Inn Albany Medical Center  ...  308.2807  228.4597
4        198232  CrestHill Suites SUNY University Albany  ...  169.3900   89.3900

[5 rows x 11 columns]

In [4]:
df.shape

(1631, 11)

In [5]:
# This is a built in magic ipython command that provides timing for code executed in the cell
%%timeit

# Run the haversine looping function
df['distance'] = haversine_looping(df)

1 loop, best of 5: 446 ms per loop


## Looping with iterrows()

In [6]:
%%timeit

# Haversine applied on rows via iteration
haversine_series = []
for index, row in df.iterrows():
    haversine_series.append(haversine(40.671, -73.985, row['latitude'], row['longitude']))
df['distance'] = haversine_series

10 loops, best of 5: 122 ms per loop


In [7]:
%%timeit

# Timing apply on the Haversine function
df['distance'] = df.apply(lambda row: haversine(40.671, -73.985, row['latitude'], row['longitude']), axis=1)

10 loops, best of 5: 49.8 ms per loop


In [9]:
 #!pip install line_profiler
 #%load_ext line_profiler

# Haversine applied on rows with line profiler
%lprun -f haversine df.apply(lambda row: haversine(40.671, -73.985, row['latitude'], row['longitude']), axis=1)


### We can note two things from the output.
1. Line 9 our trigonometric calcuation accounts for the bulk of our processing time, understandably so.
2. Each line is being hit 1631 times, that is because we're still **looping** so we iterate through every row in our dataframe

# Now for Vectorization!

- We've only been passing scalars to our Haversine function. Fortunately, all of the functions being used within the Haversine function, are also able to operate on arrays. 
- This makes the process of vectorizing our distance function quite simple as we can pass the entire series (columns) to it, performing all the calculations on the entire array simultaneously.

![alt text](https://github.com/rajeevratan84/datascienceforbusiness/blob/master/vectorization.png?raw=true)

In [10]:
%%timeit 

# Vectorized implementation of Haversine applied on Pandas series
df['distance'] = haversine(40.671, -73.985, df['latitude'], df['longitude'])

100 loops, best of 5: 1.99 ms per loop


In [11]:
# Run our line profiler to inspect further
%lprun -f haversine df['distance'] = haversine(40.671, -73.985, df['latitude'], df['longitude'])

## Let's use numpy vetorization instead of pandas

In [12]:
%%timeit

# Vectorized implementation of Haversine applied on NumPy arrays (note we use .values to access the numpy series)
df['distance'] = haversine(40.671, -73.985, df['latitude'].values, df['longitude'].values)

The slowest run took 6.44 times longer than the fastest. This could mean that an intermediate result is being cached.
1000 loops, best of 5: 263 µs per loop


# So we went from 609 ms using regular for loops, to 275 µs (or 0.275ms)!!! A 2000 fold increase in performance!